### PR TITLE
Test/big complex

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    public void test_squared() {
+        assertThat(TWO_I_TWO.squared()).isCloseTo(BigComplex.of(EIGHT, BigRationalConstants.ONE), EPSILON);
+    }
+
+    @Test
+    public void test_exp() {
+        // expected = e^2*(cos(2)+i*sin(2))
+        BigComplex expected = BigComplex.of(E.squared().multiply(TWO.cos()), E.squared().multiply(TWO.sin()));
+        assertThat(TWO_I_TWO.exp()).isCloseTo(expected, STRICT_EPS);
+    }
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:104]: This is not necessary in this class

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    public void test_multiply() {
+        assertThat(TWO_I_TWO.multiply(ONE_I_TWO)).isCloseTo(BigComplex.of(MINUS_TWO, SIX), EPSILON);
+    }
+
+    @Test
+    public void test_sqrt() {
+        assertThat(TWO_I_TWO.sqrt()).isCloseTo(BigComplex.of(BigRational.of(1.55377), BigRational.of(0.64359)),
+                STRICT_EPS);
+    }
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:86]: This is not necessary in this test class

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test
+    public void test_divide() {
+        assertThat(TWO_I_TWO.divide(ONE_I_TWO)).isCloseTo(BigComplex.of(BigRational.of(6 / 5), BigRational.of(-2 / 5)),
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:90]:
```suggestion
        assertThat(TWO_I_TWO.divide(ONE_I_TWO)).isCloseTo(BigComplex.of(BigRational.of(6.0 / 5.0), BigRational.of(-2.0 / 5.0)),
```

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void test_inverse() {
+        assertThat(TWO_I_TWO.inverse()).isCloseTo(BigComplex.of(ONE_OVER_FOUR, MINUS_ONE_OVER_FOUR), EPSILON);
+    }
+
+    @Test
+    public void test_inverse_zero() {
+        assertThrows(ArithmeticException.class, () -> ZERO.inverse());
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:63]: you can use a method reference here

Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:63]: and the fluent exception throw api from assertj

Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:63]: like this: https://www.baeldung.com/assertj-exception-assertion#2-using-assertthatexceptionoftype

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void test_angle() {
+        assertThat(TWO_I_TWO.angle()).isCloseTo(PI.divide(4), EPSILON);
+    }
+
+    @Test
+    public void test_negate() {
+        assertThat(TWO_I_TWO.negate()).isCloseTo(BigComplex.of(TWO, MINUS_TWO), EPSILON);
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:48]:
```suggestion
        assertThat(TWO_I_TWO.negate()).isCloseTo(BigComplex.of(MINUS_TWO, MINUS_TWO), EPSILON);
```

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import static io.github.ititus.math.number.BigRationalConstants.*;
+import static org.assertj.core.data.Offset.offset;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class BigComplexTests {
+    public static final BigComplex TWO_I_TWO = BigComplex.of(BigRational.of(2), BigRational.of(2));
+    public static final BigRational TWICE_TWO_SQRT = BigRational.of(2).sqrt().multiply(BigRational.of(2));
+    public static final Offset<BigRational> EPSILON = offset(of("0.001"));
+    public static final BigComplex ONE_I_TWO = BigComplex.of(BigRationalConstants.ONE, TWO);
+    public static final Offset<BigRational> STRICT_EPS = offset(of("0.000001"));
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:19]: please sort these constants and only use the strict_eps with strictOffset

**Review by iTitus — CHANGES_REQUESTED** [review]

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java (hunk 6 above)
Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:19]: remember to use the predefined constants, like `BigRationalConstants.TWO`

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java (hunk 1 above)
Comment (Segelzwerg) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:104]: Shall we move them? They pass, so why not keeping it?

Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:104]: They don't belong in this test. I want to keep the tests for _special_ logic (taylor series, other kind of series, newton approximation, ...) separate

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import static io.github.ititus.math.number.BigRationalConstants.*;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.assertj.core.data.Offset.strictOffset;
+
+public class BigComplexTests {
+    public static final BigComplex ONE_I_TWO = BigComplex.of(BigRationalConstants.ONE, TWO);
+    public static final BigComplex TWO_I_TWO = BigComplex.of(TWO, TWO);
+    public static final BigRational TWICE_TWO_SQRT = TWO.sqrt().multiply(TWO);
+    public static final Offset<BigRational> EPSILON = strictOffset(of("0.001"));
+    public static final Offset<BigRational> STRICT_EPS = strictOffset(of("0.000001"));
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:19]:
```suggestion

    private static final BigRational TWICE_TWO_SQRT = TWO.multiply(TWO.sqrt());

    private static final BigComplex ONE_I_TWO = BigComplex.of(BigRationalConstants.ONE, TWO);
    private static final BigComplex TWO_I_TWO = BigComplex.of(TWO, TWO);
    
    private static final Offset<BigRational> EPSILON = strictOffset(of("0.001"));
    private static final Offset<BigRational> STRICT_EPS = strictOffset(of("0.000001"));
```

Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:19]: Please use only this offset:
```java
private static final Offset<BigRational> EPS = strictOffset(ofExp(1, -32));
```
with some big number. `-32` ist just a suggestion.

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test
+    public void test_one_real() {
+        assertThat(ONE.getReal()).isOne();
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:23]:
```suggestion
        assertThat(ONE).real().isOne();
```

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void test_one_imag() {
+        assertThat(ONE.getImag()).isZero();
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:28]:
```suggestion
        assertThat(ONE).imag().isZero();
```

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void test_angle() {
+        assertThat(TWO_I_TWO.angle()).isCloseTo(PI.divide(4), EPSILON);
+    }
+
+    @Test
+    public void test_negate() {
+        assertThat(TWO_I_TWO.negate()).isCloseTo(BigComplex.of(MINUS_TWO, MINUS_TWO), EPSILON);
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:48]: Please use a static import for `BigComplex.of` and remove the static import for `BigRational.of`

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java (hunk 1 above)
Comment (iTitus) [reply src/test/java/io/github/ititus/math/number/BigComplexTests.java:104]: Please (re)move this

File: src/test/java/io/github/ititus/math/number/BigComplexTests.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test
+    public void test_divide() {
+        assertThat(TWO_I_TWO.divide(ONE_I_TWO)).isCloseTo(BigComplex.of(BigRational.of(6.0 / 5.0), BigRational.of(-2.0 / 5.0)),
+                EPSILON);
+    }
+
+    @Test
+    public void test_squared() {
+        assertThat(TWO_I_TWO.squared()).isCloseTo(BigComplex.of(BigRationalConstants.ZERO, EIGHT), EPSILON);
```
Comment (iTitus) [inline src/test/java/io/github/ititus/math/number/BigComplexTests.java:96]:
```suggestion
        assertThat(TWO_I_TWO.squared()).isCloseTo(imag(BigRationalConstants.EIGHT), EPSILON);
```

**Review by iTitus — CHANGES_REQUESTED** [review]
I will try to add a few more convenience factories for `BigComplex`

**Comment by iTitus** [comment]
I added a big amount of factory methods and `with`-ers. Try to use them if you can to reduce clutter!

**Review by iTitus — APPROVED** [review]

**Pull request merged**